### PR TITLE
database batching 

### PR DESCRIPTION
**Pull request by TalDerei**
references https://github.com/penumbra-zone/web/issues/1814

collectively yields **~8%** syncing improvement. 

Screenshot 2024-10-15 at 10 01 53 PM

Profiling indicates that the block processor performs substantially more database reads than writes. The most resource-intensive DB operations are "saveScanResult," "identifyNewAssets," and "resolveNullifiers," which account for the majority of disk I/O operations.

Currently, storage operations are executed sequentially (ie. saving transactions, metadata, etc.) in indexDB. Storage R/W blocks sync since each storage operation has an associated `await`, causing each to wait for the previous one to finish before continuing.

Those operations were refactored to use `Promise.all()`, ensuring the block processor isn't _artificially_ blocked on each `await` call. This means a concurrent execution of the promises, at the javascript level, meaning the calls will be _initiated_ at the same time.  But, the actual execution at the IndexedDB level won't be truly parallel because IndexedDB itself is single-threaded and performs DB operations one at a time. IndexedDB will handle them in a queue and execute them sequentially in an atomic transaction. 

The benefit of using `Promise.all()` here is the concurrent execution of **non-blocking** operations, reducing wait times between operations like network fetches and other async tasks. Therefore, other operations are not blocked while waiting for each individual write to complete. 

tldr; disk I/O doesn't seem to be a primary bottleneck. Still, disk I/O especially when filtered through the browser API / network stack is always going to be relatively slow.

### CODE REVIEW

File: packages/query/src/block-processor.ts [hunk 1]
```diff
@@ -263,11 +263,11 @@ export class BlockProcessor implements BlockProcessorInterface {
 
     // flushing is slow, avoid it until
     // - wasm says
-    // - every 1000th block
+    // - every 5000th block
     // - every block at tip
     const flushReasons = {
       scannerWantsFlush,
-      interval: compactBlock.height % 1000n === 0n,
+      interval: compactBlock.height % 5000n === 0n,
```
Comment (TalDerei) [inline packages/query/src/block-processor.ts:270]: Increasing the chunk size for flush operations improves performance.

File: packages/storage/src/indexed-db/updater.ts [hunk 2]
```diff
@@ -41,10 +41,19 @@ export class IbdUpdater {
...
+        .objectStore(update.table)
+        .put(update.value, update.key)
+        .then(() => {
+          this.notifySubscribers(update);
+        });
+      batchOperations.push(putOperation);
     }
+    await Promise.all(batchOperations);
 
     await tx.done;
```
Comment (grod220) [inline packages/storage/src/indexed-db/updater.ts:58]: thought: while I think the Promise.all() change is a good one, this doesn't change the atomicity of this operation. The transaction is building up the operations, but not actually committing anything until the even loop is released. The `await tx.done` tells us when that is. Perhaps you have a different understanding of this though?

File: packages/storage/src/indexed-db/updater.ts [hunk 3]
```diff
@@ -41,10 +41,19 @@ export class IbdUpdater {
...
+
+    // Batch all the updates into promises
     for (const update of updates.all) {
-      await tx.objectStore(update.table).put(update.value, update.key);
-      this.notifySubscribers(update);
+      const putOperation = tx
+        .objectStore(update.table)
+        .put(update.value, update.key)
+        .then(() => {
+          this.notifySubscribers(update);
```
Comment (grod220) [inline packages/storage/src/indexed-db/updater.ts:52]: issue: we should put this below `await tx.done` for all the updates given they are not committed yet

File: packages/query/src/block-processor.ts (hunk 1 above)
Comment (grod220) [reply packages/query/src/block-processor.ts:270]: Of all the changes in the PR, is it not this change that represents the nearlly all of the sync speed improvement?

**Review by grod220 — COMMENTED** [review]
I wonder if this PR kinda skirts around the core issue which is https://github.com/penumbra-zone/web/issues/1726. Aka, should we not convert our IDB implementation to a transaction+commit model instead?

File: packages/query/src/block-processor.ts (hunk 1 above)
Comment (TalDerei) [reply packages/query/src/block-processor.ts:270]: > This is something worth thinking about. Why not make the interval say 10k blocks or even bigger? I wonder what is the limiting principle for this.

the trade-off here is between memory pressure and disk I/O pressure. The longer the block processor waits to flush, the more updates accumulate in memory, but less frequent writes leading to lower I/O costs. 

In my testing, I there was a noticeable improvement when increasing the interval from 1k to 5k, but only a marginal change from 5k to 10k.

Comment (TalDerei) [reply packages/query/src/block-processor.ts:270]: > Also, we sync on `scannerWantsFlush`. This will be true when the fvk successfully decrypts something from a block. What if we don't do that though. Would it be bad if the user just waits until the next interval? Wonder if that would save time as well.

I had the same thought, and it was noticeably more performant, but it resulted in SCT state corruption.

Comment (TalDerei) [reply packages/query/src/block-processor.ts:270]: > Of all the changes in the PR, is it not this change that represents the nearlly all of the sync speed improvement?

Increasing the interval improved performance on its own, but combining it with these other changes enhanced it even further.

File: packages/storage/src/indexed-db/updater.ts (hunk 3 above)
Comment (TalDerei) [reply packages/storage/src/indexed-db/updater.ts:52]: shouldn't we just keep the status quo here and just handle block processor atomicity separately in https://github.com/penumbra-zone/web/issues/1726?

Comment (TalDerei) [reply packages/storage/src/indexed-db/updater.ts:52]: good call 👍

File: packages/query/src/block-processor.ts (hunk 1 above)
Comment (grod220) [reply packages/query/src/block-processor.ts:270]: > I had the same thought, and it was noticeably more performant, but resulted in SCT state corruption. I can investigate further.

That feels like it should not be the case 🤔 . Can you create an issue to follow up investigating why we can't flush on intervals but are required to flush on finding new notes/swaps?

**Review by grod220 — APPROVED** [review]

File: packages/storage/src/indexed-db/updater.ts [hunk 4]
```diff
@@ -41,15 +41,12 @@ export class IbdUpdater {
     const tables = updates.all.map(u => u.table);
     const tx = this.db.transaction(tables, 'readwrite');
 
-    const batchOperations = [];
-
     // Batch all the updates into promises
     for (const update of updates.all) {
-      const putOperation = tx.objectStore(update.table).put(update.value, update.key);
-      batchOperations.push(putOperation);
+      tx.objectStore(update.table).put(update.value, update.key);
```
Comment (TalDerei) [inline packages/storage/src/indexed-db/updater.ts:46]: Gabe rightly identified that performing an await inside IndexedDB transaction can risk closing the transaction prematurely in the event loop, which we shouldn't do.
